sns/parity_math.c

```diff
@@ -386,17 +746,84 @@ static void xor_calculate(struct m0_parity_math *math,
 		((uint8_t*)parity[0].b_addr)[ei] = pe;
 	}
 
+	return M0_RC(0);
 }
 
-static void reed_solomon_diff(struct m0_parity_math *math,
-			      struct m0_buf         *old,
-			      struct m0_buf         *new,
-			      struct m0_buf         *parity,
-			      uint32_t               index)
+#if ISAL_ENCODE_ENABLED
+static int isal_diff(struct m0_parity_math *math,
+		     struct m0_buf         *old,
+		     struct m0_buf         *new,
+		     struct m0_buf         *parity,
+		     uint32_t               index)
+{
+	struct m0_buf	diff_data_buf;
+	uint8_t	       *diff_data = NULL;
+	uint32_t	block_size;
+	uint32_t	alignment = sizeof(uint_fast32_t);
+	uint32_t	i;
+	int		ret = 0;
+
+	M0_ENTRY("math=%p, old=%p, new=%p, parity=%p, index=%u",
+		 math, old, new, parity, index);
+
+	M0_PRE(parity_math_invariant(math));
+	M0_PRE(old    != NULL);
+	M0_PRE(new    != NULL);
+	M0_PRE(parity != NULL);
+	M0_PRE(index  <  math->pmi_data_count);
+
+	if (old[index].b_nob != new[index].b_nob)
+		return M0_ERR_INFO(-EINVAL, "Data block size mismatch. "
+				   "Index=%u, Old data block size=%u, "
+				   "New data block size=%u", index,
+				   (uint32_t)old[index].b_nob,
+				   (uint32_t)new[index].b_nob);
+
+	block_size = new[index].b_nob;
+
+	/* It is assumed that the buffer size will always be a multiple of
+	 * 8-bytes (especially since block size currently is 4K) and this assert
+	 * is added with the hope that any deviation from this assumption is
+	 * caught during development instead of on the field.
+	 */
+	M0_ASSERT_INFO(m0_is_aligned(block_size, alignment),
+		       "block_size=%u is not %u-bytes aligned",
+		       block_size, alignment);
+
+	ALLOC_ARR_INFO(diff_data, block_size, "differential data block", ret);
+	if (diff_data == NULL)
+		return ret;
```

Comment:
Why xor_diff() and reed_solomon_diff() do not need dynamic allocations, but isal_diff() does?

Reply:
Dynamic memory allocation in isa_* functions may fail.
That's why we change these functions to return "int" to indicate error.

But here, I am not answering why dynamic memory allocation is needed.

Reply:
Here we are allocating memory diff_data (maybe I need to rename it to diff_data_arr) because we are storing differential data in this buffer.
This buffer is then passed to isal_encode_data_update() function. This function calls ec_encode_data_update() Intel ISA API to update the parity. This API needs the diff_data buffer as input to calculate/update the parity buffer.
The Intel ISA APIs determines what instruction sets are enabled and selects the appropriate version at runtime. So there may be multibyte operations depending on vectorization support.

Reply:
So, the question may be:
Can we allocate the memory in init() function?
Such that, we don't have to allocate and free the memory each time for diff/recover/encode/decode.